### PR TITLE
Add user profile feature

### PR DESCRIPTION
**Pull request by okucheruk**
This PR add user profile feature.

**Hightlights:**
- `UserProfileBuilder` is used to build immutable `UserProfile` instances.
- `UserProfileService` implements` getUser()` and `updateUser()` methods.

### CODE REVIEW

File: src/main/java/io/javaclasses/buynbye/UserProfileBuilder.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    private String deliveryAddress = "";
+
+    /**
+     * Sets login name for user profile.
+     *
+     * @param loginName login name.
+     * @return this builder instance.
+     */
+    public UserProfileBuilder withLoginName(final String loginName) {
+        checkNotNull(loginName);
```
Comment (Oleh-Stefanovych) [inline src/main/java/io/javaclasses/buynbye/UserProfileBuilder.java]: Let's show error message with `NullPointerException`

File: src/main/java/io/javaclasses/buynbye/package-info.java [hunk 2]
```diff
@@ -0,0 +1,4 @@
+/**
+ * A package which provides an API for interacting with the application server.
+ */
+package io.javaclasses.buynbye;
```
Comment (Oleh-Stefanovych) [inline src/main/java/io/javaclasses/buynbye/package-info.java]: Please add blank line at the end

File: src/test/java/io/javaclasses/buynbye/UserProfileBuilderTest.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+import java.util.stream.Stream;
+
+import static org.junit.Assert.fail;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+class UserProfileBuilderTest {
+
+    private static Stream<Arguments> provideUserProfileWithEmptyParameter() {
+        return Stream.of(
+                Arguments.of("", "Oleksii Kucheruk", "+380664811998", "Somewhere", "Name is empty")
```
Comment (Oleh-Stefanovych) [inline src/test/java/io/javaclasses/buynbye/UserProfileBuilderTest.java]: Why do we use parameterized test for the single test case?

File: src/test/java/io/javaclasses/buynbye/UserProfileServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+                        .withPhoneNumber("")
+                        .withDeliveryAddress("Same old address")
+                        .build();
+
+        UserProfile updatedJohnDoe = userProfileService.updateUser(newJohnDoe);
+
+        assertEquals(newJohnDoe.name(), updatedJohnDoe.name(),
+                "UserProfile updated incorrectly");
+
+        assertEquals("John Doe", updatedJohnDoe.name(),
```
Comment (Oleh-Stefanovych) [inline src/test/java/io/javaclasses/buynbye/UserProfileServiceTest.java]: Why do we have two asserts?

File: src/main/java/io/javaclasses/buynbye/package-info.java (hunk 2 above)
Comment (okucheruk) [reply src/main/java/io/javaclasses/buynbye/package-info.java]: Fixed.

File: src/main/java/io/javaclasses/buynbye/UserProfile.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    private String phoneNumber;
+    private String deliveryAddress;
+
+    /**
+     * Initializes a UserProfile instance.
+     *
+     * @param loginName       login name.
+     * @param name            name.
+     * @param phoneNumber     phone number.
+     * @param deliveryAddress delivery address.
```
Comment (Oleh-Stefanovych) [inline src/main/java/io/javaclasses/buynbye/UserProfile.java]: Let's use `Spine` code style configuration

File: src/main/java/io/javaclasses/buynbye/UserProfileBuilder.java (hunk 1 above)
Comment (okucheruk) [reply src/main/java/io/javaclasses/buynbye/UserProfileBuilder.java]: I think error message here will be redundant - `checkNotNull` validates only `Null` value against `checkArgument` which validates any exxpressions.

File: src/test/java/io/javaclasses/buynbye/UserProfileBuilderTest.java (hunk 3 above)
Comment (okucheruk) [reply src/test/java/io/javaclasses/buynbye/UserProfileBuilderTest.java]: This is a skeleton for checking multiple empty fields. For now only loginName field is mandatory.

File: src/test/java/io/javaclasses/buynbye/UserProfileServiceTest.java (hunk 4 above)
Comment (okucheruk) [reply src/test/java/io/javaclasses/buynbye/UserProfileServiceTest.java]: Refactored.

File: src/main/java/io/javaclasses/buynbye/UserProfileBuilder.java (hunk 1 above)
Comment (okucheruk) [reply src/main/java/io/javaclasses/buynbye/UserProfileBuilder.java]: This approach is copied from Spine.

Comment (okucheruk) [reply src/main/java/io/javaclasses/buynbye/UserProfileBuilder.java]: Agreed to use Spine approach.

File: src/test/java/io/javaclasses/buynbye/UserProfileBuilderTest.java (hunk 3 above)
Comment (okucheruk) [reply src/test/java/io/javaclasses/buynbye/UserProfileBuilderTest.java]: Test cases extended.

**Comment by okucheruk** [comment]
@Oleh-Stefanovych, fixed, PTAL.

File: src/test/java/io/javaclasses/buynbye/UserProfileBuilderTest.java [hunk 6]
```diff
@@ -16,7 +16,9 @@
 
     private static Stream<Arguments> provideUserProfileWithEmptyParameter() {
         return Stream.of(
-                Arguments.of("", "Oleksii Kucheruk", "+380664811998", "Somewhere", "Name is empty")
+                Arguments.of("", "Oleksii Kucheruk", "+380664811998", "Somewhere", "Name is empty"),
+                Arguments.of("", "", "+380664811998", "Somewhere", "Name is empty"),
+                Arguments.of("", "", "", "Somewhere", "Name is empty")
                 // Arguments.of("oleksii.kucheruk@gmail.com", "Oleksii Kucheruk", "", "Somewhere", "Phone number is empty")
```
Comment (Oleh-Stefanovych) [inline src/test/java/io/javaclasses/buynbye/UserProfileBuilderTest.java]: Please remove.

File: src/test/java/io/javaclasses/buynbye/UserProfileBuilderTest.java [hunk 7]
```diff
@@ -16,7 +16,9 @@
 
     private static Stream<Arguments> provideUserProfileWithEmptyParameter() {
         return Stream.of(
-                Arguments.of("", "Oleksii Kucheruk", "+380664811998", "Somewhere", "Name is empty")
+                Arguments.of("", "Oleksii Kucheruk", "+380664811998", "Somewhere", "Name is empty"),
+                Arguments.of("", "", "+380664811998", "Somewhere", "Name is empty"),
```
Comment (Oleh-Stefanovych) [inline src/test/java/io/javaclasses/buynbye/UserProfileBuilderTest.java]: Please extract positive cases in separate test.

**Pull request merged**